grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -340,63 +334,46 @@ def _CreateSchemas(self) -> None:
         else:
           self._CreateSchema(result_type, visiting)
 
-  def _GetSchemaOrReferenceObject(
+  def _GetReferenceObject(
       self,
       type_name: str,
       is_array: bool = False
-  ) -> Union[PrimitiveSchema, SchemaReference, ArraySchema]:
-    """Get existing `Schema Object` if primitive type, else `Reference Object`.
+  ) -> Union[SchemaReference, ArraySchema]:
+    """Get a `Reference Object` that points to a schema definition.
 
-    Primitive, not composite types don't have an actual schema, but rather an
-    equivalent OpenAPI representation that gets returned for them.
-    More complex types are expected to have been previously defined as OpenAPI
-    components and are used through OpenAPI references.
+    All types (including protobuf primitives) are expected to have been
+    previously defined in the `components` field of the root `OpenAPI Object`
+    and are used via OpenAPI references.
 
     Args:
-      type_name: The name of the type whose OpenAPI (previously constructed)
-        schema will be returned or the name of the type for which an OpenAPI
-        reference object will be created and returned.
-      is_array: A boolean flag indicating whether the selected type's schema or
-        reference object should be wrapped in an OpenAPI array as the items
-        type.
+      type_name: The name of the type for which an OpenAPI `Reference Object`
+        will be created and returned.
+      is_array: A boolean flag indicating whether the selected type's reference
+        object should be wrapped in an OpenAPI array as the items type.
 
     Returns:
-      If the `is_array` argument is set to `False`, then:
-      - if the type is primitive, returns the associated OpenAPI
-        `Schema Object`;
-      - else, an OpenAPI `Reference Object` representing the path to the actual
-        OpenAPI schema definition of the selected type.
+      If the `is_array` argument is set to `False`, then an OpenAPI `Reference
+      Object` representing the path to the actual OpenAPI schema definition of
+      the selected type.
       If the `is_array` argument is set to `True`, then an OpenAPI array schema
-      is constructed that uses for its `items` field:
-      - if the type is primitive, the actual OpenAPI schema of the type;
-      - else, an OpenAPI `Reference Object` associated with the type's schema.
+      is constructed that uses for the `items` field an OpenAPI `Reference
+      Object` associated with the type's schema.
```

Comment:
Something for the future PR: this should be probably split into two method.

> Boolean arguments loudly declare that the function does more than one thing. They are confusing and should be eliminated.
>
> — "Clean Code"

This will also allow you to avoid this nasty `Union` in the return type.

Reply:
Thank you for pointing this out! I'll make the change to 2 methods.